### DISCUSSION TITLE
Where should auth credentials be stored (e.g oauth2 access token)?

### DISCUSSION THREAD

**Discussion by giorgosera** [serverlessworkflow/specification#675] [Q&A] [answered]
Hey there, I'm a bit confused about `authDefs` and `secrets` and where should I keep auth credentials for function invocation. 

Here's my use case:

- I need to run multiple actions in a workflow. For my use case actions represent function invocations of 3rd party app APIs such as slack, typeform etc.
- For each of those services I have already went through the auth process and acquired the individual access tokens to call those APIs
- So let's say I have two actions one for invoking a 3rd party API A and another one for3rd party API B. 

Based on my current understanding there are two options for storing the auth credentials with Serverless DSL:
- Authdef: but authdef doesn't seem to have a place for an access token to be stored (unless this is what `subjectToken` is)
- Secrets: use workflow secrets and reference them functionDef arguments. As the documentation says these can be used to store OAuth tokens. 

**Questions:**

1. Which of these options is valid? Or both are invalid and there's another way which I'm missing?
2. Also, if secrets are the correct option then what is auth def used for?

I can see that there was another discussion on this and someonee suggested [here](https://github.com/serverlessworkflow/specification/issues/612) to separate `credentials` and `auth properties`.

This suggestion looks very close to what I'm trying to achieve. Any chance this is part of the spec now? 

Thanks

**Comment by ricardozanini** [comment]
Hi @giorgosera!

You can use `secrets` to store any value and retrieve it via expressions to call your functions later in the workflow.

In the OAuth2 flow, if I understood your question correctly, to me it seems that your underneath implementation should use the `auth` properties to acquire the token and pass it to the function as it calls it.

In the OpenAPI definition, you will see that each operation describes how to properly call it security-wise. So the workflow author shouldn't need to define yet another argument to call the function. Every information needed is supposed to be there.

An example:

```json
{
     "functions": [{ "operation": "withauthopenapi.yaml#securedOperation", "name": "securedFunction", 
     "authRef": { // OAuth2 definition, with the token via $SECRETS example:
         "username": "alice",
         "password": "{ $SECRET.mypassword }"
     }
      }],
     "states": [{
         "type": "operation",
         "actions": [
         {
           "functionRef": {
             "refName": "securedFunction",
             "arguments": {
               "arg1": "${ .arg1 }"
             }
          }
        }
     }]
}
```

The [`authRef`](https://github.com/serverlessworkflow/specification/blob/main/specification.md#auth-definition) will [replace any OpenAPI file definition](https://spec.openapis.org/oas/v3.1.0#oauth-flows-object), so you can override the way you call your functions there. If it's not defined, I assume the underneath implementation should take what is described in the OpenAPI as the source of truth.

The `authRef` for function calls is a new feature in the yet-to-be-released version 0.9, so the OpenAPI definition should have this info and won't be overrided by the workflow definition, nor use the secrets.

**Reply by giorgosera** [reply]
Hi @ricardozanini and thanks for your response. I think it sheds some light into my issue.

I should mention that my functions are not of type `rest` but custom. Does OpenAPI apply in that case? 
Allo, I'm calling 3rd party APIs with potentially no OpenAPI spec. So where should the OpenAPI spec be used in my case? I might be missing something very fundamental here so I apologize if this sounds to trivial. 

So if I got your reply correctly the best place to keep access tokens is in the $SECRETS array and somehow the OpenAPI spec is involved in referencing these secrets (so I won't have to pass them as arguments in the function). 

Also, in my use case the user authorizes my app for multiple 3rd party services during workflow CREATION time. Then I store all the access tokens securely before the workflow has even started. I'm trying to impersonate the user and act on their behalf so I cannot use client credential grants for example. So my problem is not so much how to get the access tokens (it's handled beforehand) but how can I reference these tokens in the DSL so that I can replace it by the correct value during runtime. 

It'd be great if `authDef` had a way for me to reference the access token (e.g. using an expression with $SECRETS.access_token).

**Reply by ricardozanini** [reply]
> how can I reference these tokens in the DSL so that I can replace it by the correct value during runtime

You can use `authDef` and `$SECRETS`. But I wouldn't since it should not be the workflow author responsible for dealing with credentials. Since you already got the tokens, your implementation should know how to map the functions to the 3rd party services to do the right call and enhance the requests with authorization info.

About openapi, since you're not using the functions type `rest` from the spec, my comment is not valid. For OpenAPI definitions, you can look at the spec and see that they define how to make the calls to each operation properly. The security mapping between the operation and the authorization method is done there. The implementation can take care of enhancing the outgoing request with the authorization tokens.

**Reply by giorgosera** [reply]
Thanks @ricardozanini! Makes sense. 

I'm not sure I understand your comment: 
> But I wouldn't since it should not be the workflow author responsible for dealing with credentials.

Maybe it is still not clear to me what the boundaries between the workflow author and the runtime are but I'd assume that the workflow author can populate the `authRef`  for the functions and the `$SECRETS` (just the names) that are needed. Then the runtime should be responsible to replace the secrets with the values at runtime. Is this correct?

**Reply by ricardozanini** [reply]
@giorgosera, this is correct. But I'd let the underneath runtime to take the information from the openapi specification and enhance the request.

You can use the `authRef` and `SECRETS` to push them as arguments, but I think this is not a domain/business problem.

Take [this as an example](https://github.com/quarkiverse/quarkus-openapi-generator/blob/main/integration-tests/example-project/src/main/openapi/open%20weather.yaml):

```yaml
  securitySchemes:
    app_id:
      type: apiKey
      description: "API key to authorize requests. (If you don't have an API key, get one at https://openweathermap.org/. See https://idratherbewriting.com/learnapidoc/docapis_get_auth_keys.html for details.)"
      name: appid
      in: query
```

The `appid` should be added in the querystring of each request. Then, in the workflow definition if you'd call it from your actions:

```
{
     "functions": [{ "operation": "weather.yaml#CurrentWeatherData", "name": "weatherData"}],
     "states": [{
         "type": "operation",
         "actions": [
         {
           "functionRef": {
             "refName": "securedFunction",
             "arguments": {
               "appid": "${ $SECRETS.appid }"
               "arg1": "${}",
               "arg2": "${}",
             }
          }
        }
     }]
}
```

Imagine having to do this for every call?

Then, what if your runtime already knows that it will call this function and already knows about the OpenAPI specification? Then, the runtime can have an external configuration that would inject this info for you as it makes the request. You don't need to provide it every time.

I'm trying to say that even if the spec allows it (and it should), your runtime can be smart and figure that alone based on the openapi definition. Does that make more sense now?

**Reply by giorgosera** [reply]
Ok that clears up a lot of things. I was under the impression that the runtime cannot make any assumptions and that the DSL should explicitly instruct the runtime. But what you said makes sense and takes advantage of open standards like OpenAPI which I believe is the philosophy of Serverless DSL.

This makes me wonder though. Since we're using `custom` functions what is more idiomatic to SW? 
- Create our own "OpenAPI" spec resource and make it accessible to the runtime via the `operation` property?
- Or use `function.metadata` to specify ? 

Does this makes sense?

**Reply by ricardozanini** [reply]
Well, for `custom` you can do whatever you want, but I'd say to abuse the `operation` attribute if possible. If it's a custom generic function that you think others would use, why not create an extension?

See in Kogito: https://kiegroup.github.io/kogito-docs/serverlessworkflow/latest/service-orchestration/configuring-openapi-services-endpoints.html

They extended the OpenAPI operation with a supportable structure from the spec pov. :)

**Reply by giorgosera** [reply]
Wow that's fantastic! Thanks for sharing.

**Comment by cdavernas** [comment, marked as answer]
> Based on my current understanding there are two options for storing the auth credentials with Serverless DSL:
> - Authdef: but authdef doesn't seem to have a place for an access token to be stored (unless this is what subjectToken is)

If you want to store a token directly, you can:

1. Use the `bearer` authentication scheme:

```yaml
...
auth:
- name: MyBearerAuth
  scheme: bearer
  properties:
     token: 'MyJwtToken'
...
```
2. Use secrets, directly in runtime expressions ($SECRETS.MYSECRET.TOKEN), or by setting the `properties` of an AuthDef to the name of the secret that contains the expected data (i.e. `{ "token": "MyToken" }`)
3. Do it properly: use OAUTH2, with the `client_credentials` flow, or, in your case, using the [`token_exchange`](https://oauth.net/2/token-exchange/) flow to impersonate a user using its token:

```yaml
...
auth:
- name: MyOAUTH2
  scheme: oauth2
  properties:
    authority: http://myauthority.com
    grant_type: urn:ietf:params:oauth:grant-type:token-exchange
    subject_token: "USER_ACCESS_TOKEN"
...
```

**Reply by giorgosera** [reply]
thanks @cdavernas it makes sense. 

Unfortunately, not all auth servers support the exchange token grant so it'll have to be the solution with the secrets.

Appreciate everyone's help in this 🙏
